### PR TITLE
Nmpolicy main flow skeleton

### PR DESCRIPTION
**Pull request by EdDev**
This change introduces the main flow of nmpolicy, focusing on the captures resolution.

It is still considered WIP as the tests have not been added.
It is posted to allow preliminary review.

### CODE REVIEW

File: nmpolicy/internal/capture/capture.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+		lexer:    leXer,
+		parser:   parser,
+		resolver: resolver,
+	}
+}
+
+func (c Capture) Resolve(
+	capturesExpr map[string]string,
+	capturesCache map[string]types.CaptureState,
+	state []byte) (map[string]types.CaptureState, error) {
```
[the remark this answers is not in the document]
Comment (EdDev) [reply nmpolicy/internal/capture/capture.go:61]: I think I answered [here](https://github.com/nmstate/nmpolicy/pull/23#discussion_r740024521) about this one. Github makes it hard to follow comments on changed code :/

I guess that it is not clear *when* we are supposed to unmarshel the input. Some parts, like the (current) state, can be unmarshal as part of the captures processing, as no one else is suppose to touch it. But maybe you suggest to process all input unmarshelling in one place.

File: nmpolicy/internal/resolver/resolver.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ */
+
+package resolver
+
+import (
+	"github.com/nmstate/nmpolicy/nmpolicy/internal/capture"
+	"github.com/nmstate/nmpolicy/nmpolicy/types"
+)
+
+type Resolver struct{}
```
[the remark this answers is not in the document]
Comment (qinqon) [reply nmpolicy/internal/resolver/resolver.go:27]: After some dicussion we can leave without the expected constructor and pass everything to the function (we will wrap it up at lexer/parser/resolver).

File: nmpolicy/internal/capture/capture.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+import (
+	"fmt"
+
+	"github.com/nmstate/nmpolicy/nmpolicy/internal/ast"
+	"github.com/nmstate/nmpolicy/nmpolicy/internal/lexer"
+	"github.com/nmstate/nmpolicy/nmpolicy/types"
+)
+
+type CapsState = map[string]types.CaptureState
```
Comment (qinqon) [inline nmpolicy/internal/capture/capture.go:30]: Can we move all the types to `types.go` ?

What do you think about moving the interfaces to their packages and unexporting the structs ? 


```golang
package lexer

type Lexer interface ...
type lexer struct ...
```

File: nmpolicy/internal/capture/capture.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (qinqon) [reply nmpolicy/internal/capture/capture.go:61]: I think the parser is able to transparently un marshal json/yaml, but we can even detect it and do proper marshaling, if needed.

**Review by qinqon — CHANGES_REQUESTED** [review]
Let's go with the empty arts constructor so we don't complicate things, but let's at least move stuff to types.go

File: nmpolicy/internal/capture/capture.go (hunk 3 above)
Comment (EdDev) [reply nmpolicy/internal/capture/capture.go:30]: This type alias is a leftover, per your request I removed them all.

Interfaces are to be defined either at the usage or in a separate package, they should almost never be defined in the implementation package.

Also, note that `types` represent public API types, and these probably do not fit there.

**Review by qinqon — APPROVED** [review]
LGTM

**Review by AlonaKaplan — APPROVED** [review]

**Comment by EdDev** [comment]
change: Rebase
